pkg/operator/ceph/csi/operator.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2024 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"fmt"
+
+	csiopv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	v1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *ReconcileCSI) createCSIOperator(csiClusterConfig *cephclient.ClusterInfo, cluster cephv1.CephCluster) error {
+
+	logger.Info("Creating ceph-CSI operator")
+	err := r.createCSIOpCephCluster(csiClusterConfig, cluster)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create ceph-CSI operator cephCluster CR")
+	}
+
+	err = r.createCSIOpCephCSIConfCR(cluster)
+	if err != nil {
+		return errors.Wrap(err, "Failed to create ceph-CSI operator config CR")
+	}
+
+	logger.Info("Successfully created ceph-CSI operator CR's")
+	return nil
+}
+
+func (r *ReconcileCSI) createCSIOpCephCluster(csiClusterConfig *cephclient.ClusterInfo, cluster cephv1.CephCluster) error {
+	logger.Info("Creating ceph-CSI operator ceph cluster CR")
+
+	csiCephCluster := &csiopv1a1.CephCluster{}
+	csiCephCluster.Name = cluster.Name
+	csiCephCluster.Namespace = cluster.Namespace
+	if cluster.Spec.CSI.ReadAffinity.Enabled {
+		csiCephCluster.Spec.ReadAffinity.CrushLocationLabels = cluster.Spec.CSI.ReadAffinity.CrushLocationLabels
+	}
+
+	mons := sets.New[string]()
+	for _, mon := range csiClusterConfig.Monitors {
+		mons.Insert(mon.Endpoint)
+	}
+	csiCephCluster.Spec.Monitors = sets.List(mons)
+
+	err := r.client.Create(r.opManagerContext, csiCephCluster)
+	if err != nil {
+		if kerrors.IsAlreadyExists(err) {
+			err = r.client.Update(r.opManagerContext, csiCephCluster)
+			if err != nil {
+				return errors.Wrapf(err, "failed to update CSI-operator ceph cluster CR %s", csiCephCluster.Name)
+			}
```

Comment:
Seems like we need to return here after the update? Otherwise it will return an err on line 73.
```suggestion
			}
			logger.Info("Successfully updated ceph-CSI operator ceph cluster CR")
			return nil
```